Failure to archive recordings for certain targets

This is observed when attempting to archive a recording from the quarkus demo app:

```
    io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
    Caused by: java.util.concurrent.ExecutionException: java.nio.file.AccessDeniedException: /deployments/quarkus-run-jar_test_20210826T213724Z.jfr
            at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
            at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
            at io.cryostat.net.web.http.api.v1.TargetRecordingPatchSave.handle(TargetRecordingPatchSave.java:66)
            at io.cryostat.net.web.http.api.v1.TargetRecordingPatchHandler.handleAuthenticated(TargetRecordingPatchHandler.java:108)
            at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:90)
            at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:65)
            at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
            at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
            at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
            at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
            at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
            at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
            at java.base/java.lang.Thread.run(Thread.java:829)
    Caused by: java.nio.file.AccessDeniedException: /deployments/quarkus-run-jar_test_20210826T213724Z.jfr
            at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:90)
            at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
            at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
            at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)
            at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:478)
            at java.base/java.nio.file.Files.newOutputStream(Files.java:220)
            at java.base/java.nio.file.Files.copy(Files.java:3067)
            at io.cryostat.core.sys.FileSystem.copy(FileSystem.java:91)
            at io.cryostat.recordings.RecordingArchiveHelper.writeRecordingToDestination(RecordingArchiveHelper.java:343)
            at io.cryostat.recordings.RecordingArchiveHelper.lambda$saveRecording$0(RecordingArchiveHelper.java:137)
            at io.cryostat.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:136)
            at io.cryostat.recordings.RecordingArchiveHelper.saveRecording(RecordingArchiveHelper.java:130)
            ... 11 more
```

The problem appears to be that the quarkus app's alias is `/deployments/quarkus-run.jar`, which appears to be an absolute path. This needs to be sanitized when the `RecordingArchiveHelper` writes the recording stream to a `.jfr` on disk.

--- a/src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java
+++ b/src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java
@@ -42,6 +42,7 @@ import java.io.IOException;
 import java.net.SocketException;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.net.URLEncoder;
 import java.net.UnknownHostException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
@@ -326,7 +327,12 @@ public class RecordingArchiveHelper {
 
         String timestamp =
                 clock.now().truncatedTo(ChronoUnit.SECONDS).toString().replaceAll("[-:]+", "");
-        String destination = String.format("%s_%s_%s", targetName, recordingName, timestamp);
+        String destination =
+                String.format(
+                        "%s_%s_%s",
+                        URLEncoder.encode(targetName, StandardCharsets.UTF_8),
+                        URLEncoder.encode(recordingName, StandardCharsets.UTF_8),
+                        timestamp);
         // TODO byte-sized rename limit is arbitrary. Probably plenty since recordings are also
         // differentiated by second-resolution timestamp
         byte count = 1;

--- a/src/test/java/io/cryostat/recordings/RecordingArchiveHelperTest.java
+++ b/src/test/java/io/cryostat/recordings/RecordingArchiveHelperTest.java
@@ -43,6 +43,8 @@ import java.io.BufferedInputStream;
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
 import java.time.Instant;
 import java.time.temporal.ChronoUnit;
@@ -248,6 +250,84 @@ class RecordingArchiveHelperTest {
                                 "some-Alias-2_someRecording_" + timestamp + ".jfr",
                                 "target",
                                 targetId));
+        Mockito.verify(notificationBuilder).build();
+        Mockito.verify(notification).send();
+    }
+
+    @ParameterizedTest
+    @ValueSource(
+            strings = {
+                "with/fs/separators",
+                "/deployment/quarkus-run.jar",
+                "with spaces",
+            })
+    void shouldSaveRecordingWithNonFilesystemSafeAlias(String alias) throws Exception {
+        Mockito.when(
+                        targetConnectionManager.executeConnectedTask(
+                                Mockito.any(),
+                                Mockito.any(TargetConnectionManager.ConnectedTask.class)))
+                .thenAnswer(
+                        new Answer<>() {
+                            @Override
+                            public Object answer(InvocationOnMock invocation) throws Throwable {
+                                TargetConnectionManager.ConnectedTask task =
+                                        (TargetConnectionManager.ConnectedTask)
+                                                invocation.getArgument(1);
+                                return task.execute(connection);
+                            }
+                        });
+        Mockito.when(connection.getService()).thenReturn(service);
+        IRecordingDescriptor descriptor = Mockito.mock(IRecordingDescriptor.class);
+        Mockito.when(descriptor.getName()).thenReturn(recordingName);
+        Mockito.when(service.getAvailableRecordings()).thenReturn(List.of(descriptor));
+
+        ServiceRef serviceRef1 =
+                new ServiceRef(
+                        URIUtil.convert(
+                                new JMXServiceURL(
+                                        "service:jmx:rmi:///jndi/rmi://cryostat:9091/jmxrmi")),
+                        alias);
+
+        Mockito.when(platformClient.listDiscoverableServices()).thenReturn(List.of(serviceRef1));
+        Mockito.when(connection.getJMXURL())
+                .thenReturn(
+                        (new JMXServiceURL("service:jmx:rmi:///jndi/rmi://cryostat:9091/jmxrmi")));
+        Mockito.when(base32.encodeAsString(Mockito.any())).thenReturn("encodedServiceUri");
+
+        Instant now = Instant.now();
+        Mockito.when(clock.now()).thenReturn(now);
+        Mockito.when(fs.exists(Mockito.any())).thenReturn(true).thenReturn(false);
+        InputStream stream = new ByteArrayInputStream("someRecordingData".getBytes());
+        Mockito.when(service.openStream(descriptor, false)).thenReturn(stream);
+        Path specificRecordingsPath = Mockito.mock(Path.class);
+        Mockito.when(archivedRecordingsPath.resolve(Mockito.anyString()))
+                .thenReturn(specificRecordingsPath);
+        Path destination = Mockito.mock(Path.class);
+        Mockito.when(specificRecordingsPath.resolve(Mockito.anyString())).thenReturn(destination);
+
+        String saveName =
+                recordingArchiveHelper
+                        .saveRecording(new ConnectionDescriptor(serviceRef1), recordingName)
+                        .get();
+
+        String timestamp = now.truncatedTo(ChronoUnit.SECONDS).toString().replaceAll("[-:]+", "");
+        String recordingName =
+                URLEncoder.encode(alias.replaceAll("[\\._]+", "-"), StandardCharsets.UTF_8)
+                        + "_someRecording_"
+                        + timestamp
+                        + ".jfr";
+        MatcherAssert.assertThat(saveName, Matchers.equalTo(recordingName));
+        Mockito.verify(fs).copy(Mockito.isA(BufferedInputStream.class), Mockito.eq(destination));
+        Mockito.verify(notificationFactory).createBuilder();
+        Mockito.verify(notificationBuilder).metaCategory("RecordingArchived");
+        Mockito.verify(notificationBuilder).metaType(HttpMimeType.JSON);
+        Mockito.verify(notificationBuilder)
+                .message(
+                        Map.of(
+                                "recording",
+                                recordingName,
+                                "target",
+                                serviceRef1.getServiceUri().toString()));
         Mockito.verify(notificationBuilder).build();
         Mockito.verify(notification).send();
     }

